<h1>Extracting and Visualizing Stock Data</h1>
<h2>Description</h2>


Extracting essential data from a dataset and displaying it is a necessary part of data science; therefore individuals can make correct decisions based on the data. In this assignment, you will extract some stock data, you will then display this data in a graph.


<h2>Table of Contents</h2>
<div class="alert alert-block alert-info" style="margin-top: 20px">
    <ul>
        <li>Define a Function that Makes a Graph</li>
        <li>Question 1: Use yfinance to Extract Stock Data</li>
        <li>Question 2: Use Webscraping to Extract Tesla Revenue Data</li>
        <li>Question 3: Use yfinance to Extract Stock Data</li>
        <li>Question 4: Use Webscraping to Extract GME Revenue Data</li>
        <li>Question 5: Plot Tesla Stock Graph</li>
        <li>Question 6: Plot GameStop Stock Graph</li>
    </ul>
<p>
    Estimated Time Needed: <strong>30 min</strong></p>
</div>

<hr>


***Note***:- If you are working in IBM Cloud Watson Studio, please replace the command for installing nbformat from `!pip install nbformat==4.2.0` to simply `!pip install nbformat`


In [1]:
"""!pip install yfinance==0.1.67
!mamba install bs4==4.10.0 -y
!pip install nbformat==4.2.0"""

'!pip install yfinance==0.1.67\n!mamba install bs4==4.10.0 -y\n!pip install nbformat==4.2.0'

In [2]:
import yfinance as yf
import pandas as pd
import requests
from bs4 import BeautifulSoup
import plotly.graph_objects as go
from plotly.subplots import make_subplots
from dateutil.parser import parse
from datetime import datetime
print("imported")


imported


In Python, you can ignore warnings using the warnings module. You can use the filterwarnings function to filter or ignore specific warning messages or categories.


In [3]:
import warnings
# Ignore all warnings
warnings.filterwarnings("ignore", category=FutureWarning)

## Define Graphing Function


In this section, we define the function `make_graph`. You don't have to know how the function works, you should only care about the inputs. It takes a dataframe with stock data (dataframe must contain Date and Close columns), a dataframe with revenue data (dataframe must contain Date and Revenue columns), and the name of the stock.


In [46]:
def make_graph(stock_data, ta, stock):
    fig = make_subplots(rows=2, cols=1, shared_xaxes=True, subplot_titles=("Historical Share Price", "Historical Revenue"), vertical_spacing = .3)
    stock_data_specific = stock_data[stock_data.Date <= pd.to_datetime('2021--06-14')]
    revenue_data_specific = revenue_data[datetime.strptime(str(revenue_data.Date),"%Y") <= pd.to_datetime('2021-04-30').year]
    fig.add_trace(go.Scatter(x=pd.to_datetime(stock_data_specific.Date, infer_datetime_format=True), y=stock_data_specific.Close.astype("float"), name="Share Price"), row=1, col=1)
    fig.add_trace(go.Scatter(x=pd.to_datetime(revenue_data_specific.Date, infer_datetime_format=True), y=revenue_data_specific.Revenue.astype("float"), name="Revenue"), row=2, col=1)
    fig.update_xaxes(title_text="Date", row=1, col=1)
    fig.update_xaxes(title_text="Date", row=2, col=1)
    fig.update_yaxes(title_text="Price ($US)", row=1, col=1)
    fig.update_yaxes(title_text="Revenue ($US Millions)", row=2, col=1)
    fig.update_layout(showlegend=False,
    height=900,
    title=stock,
    xaxis_rangeslider_visible=True)
    fig.show()

In [17]:
x=2023
y=str(x)
type(datetime.strptime(y,"%Y").year)

int

In [15]:
n=pd.to_datetime('2021-04-30').year
type(n)

int

## Question 1: Use yfinance to Extract Stock Data


Using the `Ticker` function enter the ticker symbol of the stock we want to extract data on to create a ticker object. The stock is Tesla and its ticker symbol is `TSLA`.


In [19]:
ticker=yf.Ticker("TSLA")
tesla_data=ticker.history(period="max")
tesla_data

Open        High         Low       Close  \
Date                                                                        
2010-06-29 00:00:00-04:00    1.266667    1.666667    1.169333    1.592667   
2010-06-30 00:00:00-04:00    1.719333    2.028000    1.553333    1.588667   
2010-07-01 00:00:00-04:00    1.666667    1.728000    1.351333    1.464000   
2010-07-02 00:00:00-04:00    1.533333    1.540000    1.247333    1.280000   
2010-07-06 00:00:00-04:00    1.333333    1.333333    1.055333    1.074000   
...                               ...         ...         ...         ...   
2023-11-15 00:00:00-05:00  239.289993  246.699997  236.449997  242.839996   
2023-11-16 00:00:00-05:00  239.490005  240.880005  230.960007  233.589996   
2023-11-17 00:00:00-05:00  232.000000  237.389999  226.539993  234.300003   
2023-11-20 00:00:00-05:00  234.039993  237.100006  231.020004  235.600006   
2023-11-21 00:00:00-05:00  235.039993  243.619995  233.339996  241.199997   

                              Volume  Dividends  Stock Splits  
Date                                                           
2010-06-29 00:00:00-04:00  281494500        0.0           0.0  
2010-06-30 00:00:00-04:00  257806500        0.0           0.0  
2010-07-01 00:00:00-04:00  123282000        0.0           0.0  
2010-07-02 00:00:00-04:00   77097000        0.0           0.0  
2010-07-06 00:00:00-04:00  103003500        0.0           0.0  
...                              ...        ...           ...  
2023-11-15 00:00:00-05:00  150354000        0.0           0.0  
2023-11-16 00:00:00-05:00  136816800        0.0           0.0  
2023-11-17 00:00:00-05:00  142532800        0.0           0.0  
2023-11-20 00:00:00-05:00  116320100        0.0           0.0  
2023-11-21 00:00:00-05:00  121987800        0.0           0.0  

[3374 rows x 7 columns]

Using the ticker object and the function `history` extract stock information and save it in a dataframe named `tesla_data`. Set the `period` parameter to `max` so we get information for the maximum amount of time.


In [20]:
ticker=yf.Ticker("TSLA")
tesla_data=ticker.history(period="max")
tesla_data.reset_index(inplace=True)
tesla_data

Date        Open        High         Low  \
0    2010-06-29 00:00:00-04:00    1.266667    1.666667    1.169333   
1    2010-06-30 00:00:00-04:00    1.719333    2.028000    1.553333   
2    2010-07-01 00:00:00-04:00    1.666667    1.728000    1.351333   
3    2010-07-02 00:00:00-04:00    1.533333    1.540000    1.247333   
4    2010-07-06 00:00:00-04:00    1.333333    1.333333    1.055333   
...                        ...         ...         ...         ...   
3369 2023-11-15 00:00:00-05:00  239.289993  246.699997  236.449997   
3370 2023-11-16 00:00:00-05:00  239.490005  240.880005  230.960007   
3371 2023-11-17 00:00:00-05:00  232.000000  237.389999  226.539993   
3372 2023-11-20 00:00:00-05:00  234.039993  237.100006  231.020004   
3373 2023-11-21 00:00:00-05:00  235.039993  243.619995  233.339996   

           Close     Volume  Dividends  Stock Splits  
0       1.592667  281494500        0.0           0.0  
1       1.588667  257806500        0.0           0.0  
2       1.464000  123282000        0.0           0.0  
3       1.280000   77097000        0.0           0.0  
4       1.074000  103003500        0.0           0.0  
...          ...        ...        ...           ...  
3369  242.839996  150354000        0.0           0.0  
3370  233.589996  136816800        0.0           0.0  
3371  234.300003  142532800        0.0           0.0  
3372  235.600006  116320100        0.0           0.0  
3373  241.199997  121987800        0.0           0.0  

[3374 rows x 8 columns]

**Reset the index** using the `reset_index(inplace=True)` function on the tesla_data DataFrame and display the first five rows of the `tesla_data` dataframe using the `head` function. Take a screenshot of the results and code from the beginning of Question 1 to the results below.


In [21]:
tesla_data.reset_index(inplace=True)
tesla_data.head(5)

index                      Date      Open      High       Low     Close  \
0      0 2010-06-29 00:00:00-04:00  1.266667  1.666667  1.169333  1.592667   
1      1 2010-06-30 00:00:00-04:00  1.719333  2.028000  1.553333  1.588667   
2      2 2010-07-01 00:00:00-04:00  1.666667  1.728000  1.351333  1.464000   
3      3 2010-07-02 00:00:00-04:00  1.533333  1.540000  1.247333  1.280000   
4      4 2010-07-06 00:00:00-04:00  1.333333  1.333333  1.055333  1.074000   

      Volume  Dividends  Stock Splits  
0  281494500        0.0           0.0  
1  257806500        0.0           0.0  
2  123282000        0.0           0.0  
3   77097000        0.0           0.0  
4  103003500        0.0           0.0

## Question 2: Use Webscraping to Extract Tesla Revenue Data


Use the `requests` library to download the webpage https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/revenue.htm Save the text of the response as a variable named `html_data`.


In [23]:
url="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/revenue.htm"

In [24]:
html_data=requests.get(url).text

Parse the html data using `beautiful_soup`.


Using `BeautifulSoup` or the `read_html` function extract the table with `Tesla Revenue` and store it into a dataframe named `tesla_revenue`. The dataframe should have columns `Date` and `Revenue`.


In [25]:
soup=BeautifulSoup(html_data,"html5lib")
# create a dataframe for tesla 
tesla_revenue=pd.DataFrame(columns=['Data',"Revenue"])


for row in soup.find_all("tbody")[1].find_all("tr"):
    col = row.find_all("td")
    Date=col[0].text
    Revenue=col[1].text
    #append a row to the dataframe 
    tesla_revenue.append({"Data":Date,"Revenue":Revenue},ignore_index=True)
tesla_revenue.head()


Empty DataFrame
Columns: [Data, Revenue]
Index: []

In [27]:
soup=BeautifulSoup(html_data,"html5lib")
#soup.find_all("tbody")[0].find_all("tr")

<details><summary>Click here if you need help locating the table</summary>

```
    
Below is the code to isolate the table, you will now need to loop through the rows and columns like in the previous lab
    
soup.find_all("tbody")[1]
    
If you want to use the read_html function the table is located at index 1


```

</details>


In [28]:
# we try another way 
tables_list=pd.read_html(str(html_data))
tesla_data2=pd.DataFrame(tables_list[0])
tesla_data2.columns=["Date","Revenue"]
tesla_revenue=tesla_data2
tesla_revenue

Date  Revenue
0   2021  $53,823
1   2020  $31,536
2   2019  $24,578
3   2018  $21,461
4   2017  $11,759
5   2016   $7,000
6   2015   $4,046
7   2014   $3,198
8   2013   $2,013
9   2012     $413
10  2011     $204
11  2010     $117
12  2009     $112

Execute the following line to remove the comma and dollar sign from the `Revenue` column. 


In [29]:
tesla_revenue["Revenue"] = tesla_revenue['Revenue'].str.replace(',|\$',"")
#tesla_revenue['Date']=tesla_revenue['Date'].to_pydatetime()
#for  date in tesla_revenue['Date']:
 #  date=date.to_pydatetime()
tesla_revenue

Date Revenue
0   2021   53823
1   2020   31536
2   2019   24578
3   2018   21461
4   2017   11759
5   2016    7000
6   2015    4046
7   2014    3198
8   2013    2013
9   2012     413
10  2011     204
11  2010     117
12  2009     112

Execute the following lines to remove an null or empty strings in the Revenue column.


In [30]:
tesla_revenue.dropna(inplace=True)

tesla_revenue = tesla_revenue[tesla_revenue['Revenue'] != ""]
tesla_revenue.shape

(13, 2)

Display the last 5 row of the `tesla_revenue` dataframe using the `tail` function. Take a screenshot of the results.


In [31]:
tesla_revenue.tail()

Date Revenue
8   2013    2013
9   2012     413
10  2011     204
11  2010     117
12  2009     112

## Question 3: Use yfinance to Extract Stock Data


Using the `Ticker` function enter the ticker symbol of the stock we want to extract data on to create a ticker object. The stock is GameStop and its ticker symbol is `GME`.


In [32]:
ticker=yf.Ticker("GME")


Using the ticker object and the function `history` extract stock information and save it in a dataframe named `gme_data`. Set the `period` parameter to `max` so we get information for the maximum amount of time.


In [33]:
gme_data=ticker.history(period="max")
gme_data.head()

Open      High       Low     Close    Volume  \
Date                                                                          
2002-02-13 00:00:00-05:00  1.620129  1.693350  1.603296  1.691667  76216000   
2002-02-14 00:00:00-05:00  1.712708  1.716074  1.670626  1.683251  11021600   
2002-02-15 00:00:00-05:00  1.683251  1.687459  1.658002  1.674834   8389600   
2002-02-19 00:00:00-05:00  1.666418  1.666418  1.578047  1.607504   7410400   
2002-02-20 00:00:00-05:00  1.615920  1.662210  1.603296  1.662210   6892800   

                           Dividends  Stock Splits  
Date                                                
2002-02-13 00:00:00-05:00        0.0           0.0  
2002-02-14 00:00:00-05:00        0.0           0.0  
2002-02-15 00:00:00-05:00        0.0           0.0  
2002-02-19 00:00:00-05:00        0.0           0.0  
2002-02-20 00:00:00-05:00        0.0           0.0

**Reset the index** using the `reset_index(inplace=True)` function on the gme_data DataFrame and display the first five rows of the `gme_data` dataframe using the `head` function. Take a screenshot of the results and code from the beginning of Question 3 to the results below.


In [34]:
gme_data.reset_index(inplace=True)
gme_data

Date       Open       High        Low      Close  \
0    2002-02-13 00:00:00-05:00   1.620129   1.693350   1.603296   1.691667   
1    2002-02-14 00:00:00-05:00   1.712708   1.716074   1.670626   1.683251   
2    2002-02-15 00:00:00-05:00   1.683251   1.687459   1.658002   1.674834   
3    2002-02-19 00:00:00-05:00   1.666418   1.666418   1.578047   1.607504   
4    2002-02-20 00:00:00-05:00   1.615920   1.662210   1.603296   1.662210   
...                        ...        ...        ...        ...        ...   
5477 2023-11-15 00:00:00-05:00  13.150000  13.500000  13.030000  13.270000   
5478 2023-11-16 00:00:00-05:00  13.140000  13.190000  12.520000  12.530000   
5479 2023-11-17 00:00:00-05:00  12.520000  13.060000  12.520000  13.030000   
5480 2023-11-20 00:00:00-05:00  13.000000  13.200000  12.800000  12.800000   
5481 2023-11-21 00:00:00-05:00  12.650000  12.720000  12.260000  12.550000   

        Volume  Dividends  Stock Splits  
0     76216000        0.0           0.0  
1     11021600        0.0           0.0  
2      8389600        0.0           0.0  
3      7410400        0.0           0.0  
4      6892800        0.0           0.0  
...        ...        ...           ...  
5477   4121600        0.0           0.0  
5478   3425800        0.0           0.0  
5479   3414200        0.0           0.0  
5480   2294900        0.0           0.0  
5481   3707300        0.0           0.0  

[5482 rows x 8 columns]

## Question 4: Use Webscraping to Extract GME Revenue Data


Use the `requests` library to download the webpage https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/stock.html. Save the text of the response as a variable named `html_data`.


In [35]:
url="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/stock.html"

In [37]:
html_data=requests.get(url).text

Parse the html data using `beautiful_soup`.


In [38]:
soup=BeautifulSoup(html_data,'html5lib')

Using `BeautifulSoup` or the `read_html` function extract the table with `GameStop Revenue` and store it into a dataframe named `gme_revenue`. The dataframe should have columns `Date` and `Revenue`. Make sure the comma and dollar sign is removed from the `Revenue` column using a method similar to what you did in Question 2.


<details><summary>Click here if you need help locating the table</summary>

```
    
Below is the code to isolate the table, you will now need to loop through the rows and columns like in the previous lab
    
soup.find_all("tbody")[1]
    
If you want to use the read_html function the table is located at index 1


```

</details>


In [39]:
gme_revenue=pd.DataFrame(columns=["Date","Revenue"])

In [40]:
for row in soup.find_all("tbody")[1].find_all("tr"):
    cols=row.find_all("td")
    date=cols[0].text
    rev=cols[1].text
    #append the record 
    gme_revenue.append({"Date":date,"Revenue":rev},ignore_index=True)
gme_revenue

Empty DataFrame
Columns: [Date, Revenue]
Index: []

In [41]:
# try another way
my_list=pd.read_html(url)
gme_revenue=pd.DataFrame(my_list[1])
gme_revenue.columns=["Date","Revenue"]


In [78]:
# create an empty dataframe
""""gme_revenue=pd.DataFrame(columns=["Date","Revenue"])

tables=soup.find_all("tbody")
#type(tables_list)
tables.last()
length = tables.getRow()

print(length)"""

'"gme_revenue=pd.DataFrame(columns=["Date","Revenue"])\n\ntables=soup.find_all("tbody")\n#type(tables_list)\ntables.last()\nlength = tables.getRow()\n\nprint(length)'

Display the last five rows of the `gme_revenue` dataframe using the `tail` function. Take a screenshot of the results.


In [42]:
gme_revenue.tail()

Date Revenue
57  2006-01-31  $1,667
58  2005-10-31    $534
59  2005-07-31    $416
60  2005-04-30    $475
61  2005-01-31    $709

In [27]:
type(gme_revenue['Date'][1])

str

In [ ]:
#convert str to datetime 
#strptime(str_object,day_first=)

## Question 5: Plot Tesla Stock Graph


Use the `make_graph` function to graph the Tesla Stock Data, also provide a title for the graph. The structure to call the `make_graph` function is `make_graph(tesla_data, tesla_revenue, 'Tesla')`. Note the graph will only show data upto June 2021.


In [47]:
make_graph(tesla_data, tesla_revenue, 'Tesla')

TypeError: Invalid comparison between dtype=datetime64[ns, America/New_York] and Timestamp

## Question 6: Plot GameStop Stock Graph


Use the `make_graph` function to graph the GameStop Stock Data, also provide a title for the graph. The structure to call the `make_graph` function is `make_graph(gme_data, gme_revenue, 'GameStop')`. Note the graph will only show data upto June 2021.


In [48]:
make_graph(gme_data, gme_revenue, 'GameStop')

TypeError: Invalid comparison between dtype=datetime64[ns, America/New_York] and Timestamp

<h2>About the Authors:</h2> 

<a href="https://www.linkedin.com/in/joseph-s-50398b136/">Joseph Santarcangelo</a> has a PhD in Electrical Engineering, his research focused on using machine learning, signal processing, and computer vision to determine how videos impact human cognition. Joseph has been working for IBM since he completed his PhD.

Azim Hirjani


## Change Log

| Date (YYYY-MM-DD) | Version | Changed By    | Change Description        |
| ----------------- | ------- | ------------- | ------------------------- |
| 2022-02-28        | 1.2     | Lakshmi Holla | Changed the URL of GameStop |
| 2020-11-10        | 1.1     | Malika Singla | Deleted the Optional part |
| 2020-08-27        | 1.0     | Malika Singla | Added lab to GitLab       |

<hr>

## <h3 align="center"> © IBM Corporation 2020. All rights reserved. <h3/>

<p>
